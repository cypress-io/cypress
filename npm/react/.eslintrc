{
  "plugins": [
    "cypress",
    "@cypress/dev"
  ],
  "extends": [
    "plugin:@cypress/dev/general",
    "plugin:@cypress/dev/tests",
    "plugin:@cypress/dev/react"
  ],
  "parser": "@typescript-eslint/parser",
  "env": {
    "cypress/globals": true
  },
  "root": true,
  "rules": {
    "no-console": "off",
    "mocha/no-global-tests": "off",
    "react/jsx-filename-extension": [
      "warn",
      {
        "extensions": [
          ".js",
          ".jsx",
          ".tsx"
        ]
      }
    ]
  }
}
