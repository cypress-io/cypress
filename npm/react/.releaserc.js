/**
 * NOTICE
 * Disable the release for React until further notice.
 * You may re-enable it by toggling publishConfig: restricted/public
 * and uncommenting the following lines
 */
module.exports = {
  ...require('../../.releaserc.base'),
  branches: [
    { name: 'master', channel: 'next' },
  ],
}
