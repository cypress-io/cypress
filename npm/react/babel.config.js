module.exports = {
  presets: [
    '@babel/preset-env',
    '@babel/preset-react',
    '@babel/preset-typescript',
  ],
  plugins: [
    // allow lazy loaded components with dynamic "import(...)"
    // https://babeljs.io/docs/en/babel-plugin-syntax-dynamic-import/
    '@babel/plugin-syntax-dynamic-import',
    '@babel/plugin-proposal-class-properties',
    // https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs
    // loose ES6 modules allow us to dynamically mock imports during tests
    [
      '@babel/plugin-transform-modules-commonjs',
      {
        loose: true,
      },
    ],
    [
      'module-resolver',
      {
        alias: {
          '@cypress/react': './dist/cypress-react.cjs.js',
        },
      },
    ],
  ],
}
