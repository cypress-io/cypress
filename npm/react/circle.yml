version: 2.1
orbs:
  cypress: cypress-io/cypress@1.26.0

workflows:
  build:
    jobs:
      # install and cache dependencies in this job
      # AND build the library once
      # then the workspace will be passed to other jobs
      - cypress/install:
          name: Install
          executor: cypress/base-12
          build: npm run transpile
          post-steps:
            - run:
                name: Show info üì∫
                command: npx cypress info
            - run:
                name: Linting code üßπ
                command: npm run lint
            - run:
                name: Stop exclusive tests 1Ô∏è‚É£
                command: npm run stop-only
            - run:
                name: Build folder üèó
                command: npm run build
            - run:
                name: Run unit tests üë∑
                command: npm run test:unit

      - cypress/run:
          name: Example A11y
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/a11y
          command: npm test
          store_artifacts: true

      - cypress/run:
          name: Example Babel
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: false
          working_directory: examples/using-babel
          command: npm test
          store_artifacts: true

      - cypress/run:
          name: Example Babel + Typescript
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/using-babel-typescript
          command: npm test
          store_artifacts: true

      - cypress/run:
          name: Example React Scripts
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/react-scripts
          command: npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/react-scripts

      - cypress/run:
          name: Example Next.js
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/nextjs
          command: npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/nextjs

      - cypress/run:
          # react-scripts example with component tests not in "src" folder
          # but in "cypress/component" folder
          name: Example Component Folder
          executor: cypress/base-12
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/react-scripts-folder
          command: npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/react-scripts-folder

      - cypress/run:
          name: Example Tailwind
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          executor: cypress/base-12
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/tailwind
          command: |
            DEBUG=cypress-react-unit-test,find-webpack npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  ls -la
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/tailwind

      - cypress/run:
          name: Example Webpack file
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          executor: cypress/base-12
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/webpack-file
          command: npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/webpack-file

      - cypress/run:
          name: Example Rollup
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          executor: cypress/base-12
          install-command: npm install --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/rollup
          command: npm test
          store_artifacts: true

      - cypress/run:
          name: Example Webpack options
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          executor: cypress/base-12
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/webpack-options
          command: npm test
          store_artifacts: true
          post-steps:
            - run:
                name: Check coverage üìà
                command: |
                  npm run check-coverage
                  npm run only-covered
                working_directory: examples/webpack-options

#      - cypress/run:
#          name: Example Sass
#          requires:
#            - Install
#          # we need the same OS version as in install job
#          # because we will use native Sass dependency
#          executor: cypress/base-12
#          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
#          install-command: npm install  --no-bin-links
#          verify-command: echo 'Already verified'
#          no-workspace: true
#          working_directory: examples/sass-and-ts
#          command: npm test
#          store_artifacts: true
#          post-steps:
#            - run:
#                name: Check coverage üìà
#                command: |
#                  npm run check-coverage
#                  npm run only-covered
#                working_directory: examples/sass-and-ts

      - cypress/run:
          name: Example Snapshots
          requires:
            - Install
          executor: cypress/base-12
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/snapshots
          command: npm test
          store_artifacts: true

      - cypress/run:
          name: Visual Sudoku
          executor: cypress/base-12
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/visual-sudoku
          command: npm test
          store_artifacts: true
          post-steps:
            - store_artifacts:
                path: examples/visual-sudoku/cypress/snapshots

      - cypress/run:
          name: Visual with Applitools
          executor: cypress/base-12
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/visual-testing-with-applitools
          # to correctly run this job, we need Applitools token
          # external pull requests do not have environment variables set
          # thus the job will always fail. Let's skip this job if the
          # environment variable is missing
          command: |
            if [ -z "$APPLITOOLS_API_KEY" ]; then
              echo "Skipping Applitools test job, missing environment variable APPLITOOLS_API_KEY"
            else
              npm test
            fi
          store_artifacts: true

      - cypress/run:
          name: Visual with Percy
          executor: cypress/base-12
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/visual-testing-with-percy
          # run Percy agent and then run Cypress component tests
          # https://docs.percy.io/docs/cypress
          command: npx percy exec -- npm test
          store_artifacts: true

      - cypress/run:
          name: Visual with Happo
          executor: cypress/base-12
          requires:
            - Install
          # each example installs "cypress-react-unit-test" as a local dependency (symlink)
          install-command: npm install  --no-bin-links
          verify-command: echo 'Already verified'
          no-workspace: true
          working_directory: examples/visual-testing-with-happo
          command: npm run test:happo
          store_artifacts: true

      - cypress/run:
          name: Component Tests
          executor: cypress/base-12
          parallelism: 4
          requires:
            - Install
          # notice a trick to avoid re-installing dependencies
          # in this job - a do-nothing "install-command" parameter
          install-command: echo 'Nothing to install in this job'
          # we are not going to use results from this job anywhere else
          no-workspace: true
          record: false
          store_artifacts: true
          # following examples from
          # https://circleci.com/docs/2.0/parallelism-faster-jobs/
          # TODO probably only run component tests and move integration sanity checks into own job
          command: |
            TESTFILES=$(circleci tests glob "cypress/{component,integration}/**/*spec.{js,jsx,ts,tsx}" | circleci tests split --total=4)
            echo "Test files for this machine are $TESTFILES"
            npx cypress run --spec $TESTFILES

      # this job attaches the workspace left by the install job
      # so it is ready to run Cypress tests
      # only we will run semantic release script instead
      - cypress/run:
          name: NPM release
          # only run NPM release from specific branch(es)
          filters:
            branches:
              only:
                - main
          # we need newer Node for semantic release
          executor: cypress/base-12
          requires:
            - Install
            - Component Tests
            - Example A11y
            - Example Babel
            - Example Component Folder
            - Example React Scripts
            # - Example Sass
            - Example Snapshots
            - Example Tailwind
            - Example Webpack file
            - Example Webpack options
            - Example Rollup
            - Visual Sudoku
            - Visual with Percy
            - Visual with Happo
            - Visual with Applitools
          install-command: echo 'Already installed'
          verify-command: echo 'Already verified'
          no-workspace: true
          # instead of "cypress run" do NPM release üòÅ
          # clear environment variables to trick semantic-release
          # into thinking this is NOT a pull request.
          # (under the hood the module env-ci is used to check if this is a PR)
          command: |
            npm run build
            CIRCLE_PR_NUMBER= \
            CIRCLE_PULL_REQUEST= \
            CI_PULL_REQUEST= \
            npm run semantic-release
