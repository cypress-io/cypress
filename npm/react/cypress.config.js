module.exports = {
  'viewportWidth': 400,
  'viewportHeight': 400,
  'video': false,
  'projectId': 'z9dxah',
  'env': {
    'reactDevtools': true,
  },
  'experimentalFetchPolyfill': true,
  'component': {
    'excludeSpecPattern': [
      '**/__snapshots__/*',
      '**/__image_snapshots__/*',
      'examples/**/*',
    ],
    devServer: {
      framework: 'react',
      bundler: 'vite',
    },
  },
}
