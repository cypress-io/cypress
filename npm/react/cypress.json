{
  "viewportWidth": 400,
  "viewportHeight": 400,
  "video": false,
  "projectId": "z9dxah",
  "testFiles": "**/*spec.{js,jsx,ts,tsx}",
  "env": {
    "cypress-react-selector": {
      "root": "#cypress-root"
    }
  },
  "ignoreTestFiles": [
    "**/__snapshots__/*",
    "**/__image_snapshots__/*"
  ],
  "experimentalComponentTesting": true,
  "experimentalFetchPolyfill": true
}
