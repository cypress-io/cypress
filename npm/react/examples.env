
/examples/nextjs
# /examples/nextjs-webpack-5
/examples/react-scripts
/examples/react-scripts-typescript
/examples/find-webpack
/examples/webpack-file
/examples/react-scripts-folder
/examples/using-babel-typescript
/examples/webpack-options
/examples/sass-and-ts
