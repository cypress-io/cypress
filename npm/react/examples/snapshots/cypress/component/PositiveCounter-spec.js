/// <reference types="cypress" />
import React from 'react'
import { mount } from '@cypress/react'
import PositiveCounter from './PositiveCounter'

// NOTE: current version 1.4.3 of cypress-plugin-snapshots only works with
// objects and images.
// https://github.com/meinaart/cypress-plugin-snapshots/issues/122
it.skip('sanity check: string snapshots work', () => {
  cy.wrap('hello, world').toMatchSnapshot()
})

// utility child command that grabs element's HTML and snapshots its as an object
Cypress.Commands.add('toMatchHTML', { prevSubject: 'element' }, ($el) => {
  return cy
  .wrap({
    html: $el[0].outerHTML,
  })
  .toMatchSnapshot()
})

describe('PositiveCounter', () => {
  it('starts with zero', () => {
    mount(<PositiveCounter />)
    cy.contains('Value: 0')
    // previous command yields jQuery element
    // I would like to get its outer HTML which
    // we can do via $el[0].outerHTML shorthand
    .its('0.outerHTML')
    // convert text to JSON object for the snapshot plugin to work
    // https://github.com/meinaart/cypress-plugin-snapshots/issues/122
    .then((html) => {
      return {
        html,
      }
    })
    .toMatchSnapshot()

    // in other tests we will use utility child command .toMatchHTML()
  })

  it('should render counts', () => {
    mount(<PositiveCounter />)
    cy.get('.increment')
    .click()
    .click()
    .click()

    // make sure the component updates
    cy.contains('Value: 3').toMatchHTML()

    cy.get('.increment')
    .click()
    .click()
    .click()

    cy.contains('Value: 6').toMatchHTML()
  })

  it('should not go negative', () => {
    mount(<PositiveCounter />)
    cy.get('.increment').click()
    cy.get('.decrement')
    .click()
    .click()

    cy.contains('Value: 0').toMatchHTML()
  })

  it('snapshots the component state', () => {
    mount(<PositiveCounter />)
    cy.get('.increment')
    .click()
    .click()
    .click()
    .click()

    // The component's code set its reference
    // as a property on the "window" object
    // when running inside Cypress. This allows
    // the test to access it.
    cy.window()
    .its('PositiveCounter.state')
    .toMatchSnapshot()
  })
})
