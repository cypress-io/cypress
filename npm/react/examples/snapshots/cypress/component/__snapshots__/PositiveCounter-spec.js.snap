exports[`PositiveCounter > should render counts #0`] =
{
  "html": "<span>Value: 3<button class=\"decrement\">−</button><button class=\"increment\">+</button></span>"
};

exports[`PositiveCounter > should render counts #1`] =
{
  "html": "<span>Value: 6<button class=\"decrement\">−</button><button class=\"increment\">+</button></span>"
};

exports[`PositiveCounter > should not go negative #0`] =
{
  "html": "<span>Value: 0<button class=\"decrement\">−</button><button class=\"increment\">+</button></span>"
};

exports[`PositiveCounter > snapshots the component state #0`] =
{
  "count": 4
};

exports[`PositiveCounter > starts with zero #0`] =
{
  "html": "<span>Value: 0<button class=\"decrement\">−</button><button class=\"increment\">+</button></span>"
};
